### ISSUE TITLE
Dashboard not updated correctly if two people have the same RSS feed 

### ISSUE THREAD

**Issue by ftisiot** [forem/forem#13185]
If two people share the same `RSS Feed URL` in `Settings` -> `Publishing to DEV Community from RSS`
Then new post will appear randomly only in 1 feed.

1. Have two separated dev.to accounts
2. Go to `Settings` -> `Publishing to DEV Community from RSS` and setup the same `RSS Feed URL` in both
3. Create a new RSS entry
4. The entry is show in one of the two dev.to account's (say account A) Dashboard and not in the other (account B)
5. If the account who doesn't see the RSS feed (account B) tries creating a new post mentioning the new entry's canonical_url it gets a message 
`canonical_url: has already been taken`
6. If the account who does see the RSS entry (account A) deletes the draft, then Account B can publish successfully the content

**Expected behavior**

Both dev.to accounts should see the new RSS entry in the dashboard

**Screenshots**

**Desktop (please complete the following information):**

- OS, version: MacOS Big Sur 11.2.3
- Browser, version: tried with Safari, Chrome and Firefox

**Smartphone (please complete the following information):**

- Device:
- OS, version:
- Browser, version:

**Additional context**

**Comment by rhymes** [comment]
I think this is a variation of https://github.com/forem/forem/issues/1845

**Comment by squarebat** [comment]
I'd like to take this up

**Comment by rhymes** [comment]
Thanks @squarebat, let us know if you need help :)

**Comment by squarebat** [comment]
Thanks @rhymes. The solution seems straight forward, but the real issue is with the unique `canonical_url`. Could I get rid of that constraint to allow multiple posts to have the same canonical_url or do we need it to be unique?

**Comment by rhymes** [comment]
@squarebat In my opinion only one article should be the canonical URL, at all times, because that's how we signal search engines which is the authoritative version of an article. 

We have had some UX related issues like people from the same org importing the feed but one of them failing to publish because someone else already published with the canonical URL.

In https://github.com/forem/forem/pull/10161#issuecomment-685902405 both @mstruve and @citizen428 validated the approach of having a global canonical URL, regardless of publication status.

If we want to re-visit the possibility of having `canonical_url, published` as a unique combination (so that multiple people can import the same feed), we could revisit [this comment](https://github.com/forem/forem/pull/9788/files#r480880463) by @citizen428, using a partial unique index, something like: `ON articles (canonical_url, published) WHERE published`.

This would allow you to have as many canonical urls that are the same value in drafts and then only one that's published.

With a corresponding Rails validation I guess.

Hope this helps!

**Comment by squarebat** [comment]
Thanks @rhymes! That helps a lot!

**Comment by squarebat** [comment]
tl;dr : I played around with two dev accounts and based on my observations, some issues will occur regardless of whether we fix this issue or not. My observations might be incorrect, correct me if I'm wrong.

Going through the PRs mentioned by @rhymes helped a lot. It's quite obvious now why canonical_url needs to be unique. I played around with importing from the same RSS feed from two dev accounts and this is what I observed:
- If an article from the feed was already imported in one account A, it won't be imported in account B regardless of whether the canonical_url was set or not, or if it was draft/published. 
- If the article is deleted from account A, it can be fetched in account B.
 
Note: These observations are from using the `Fetch feed now` button, and not from periodic feed imports.
I can think of following issues that will occur:

**If we allow both accounts to fetch the same article:**
- None of them gets to publish the post if the canonical_url is same. They can be published by changing it, but since there is no way to know which account is using that it, the canonical_url might remain unpublished. 

**If we allow only one account to fetch an article (as is the case right now):**
- There's a UX issue where user won't know why articles aren't being fetched from the feed.
- Articles are being fetched on first come first serve basis, which defeats the purpose of allowing 2 accounts to import from the same feed.

**Comment by citizen428** [comment]
@squarebat Thanks for this nice write-up. I think it may be better to tackle this one level higher up, e.g. maybe two users should be able to use the same feed URL? I haven't thought through all product considerations here but I think most use cases where that could make sense are probably better handled by creating an organization.

**Comment by squarebat** [comment]
@citizen428 Yeah, it's definitely better to let multiple accounts access the same feed. Publishing articles from feeds with multiple access can be managed easily by creating an organization, however any user not part of the organization that accesses the feed will create conflicts. From a product perspective, this will happen rarely, though it has caused issues in the past as detailed in #12708 .

**Comment by rhymes** [comment]
Hi @squarebat, great analysis! I think we should allow two people to fetch the same feed, so we definitely need to alter that logic that makes it impossible for a second person to import an article that's already present in the DB. Secondly, can we alter the canonical url validation, so that the UX tells the user who is the user that owns the `canonical_url` for that article? I need confirmation from @forem/product on this (there might be privacy issues) but overall it feels like the only practical solution

**Comment by squarebat** [comment]
@rhymes exactly what I was thinking. We should let users know about other users accessing the feed. Though in another issue someone mentioned it might lead to harassment. However, I don't see a win-win situation here and this feels like the best possible solution. Besides, harassment can be dealt with by reporting to the Forem team.

I'll create a draft PR, solving the article multiple fetch issue. Then we can figure out how to move ahead.

**Comment by citizen428** [comment]
@rhymes @squarebat To avoid privacy issues maybe we could automatically flag this to admins with all the necessary info. So instead of showing the user an error that says something like "Can't import this article because it is already owned by user X" it says something like "Can't import this article our admins are looking into this. If you want to provide more detail email ....".

**Comment by squarebat** [comment]
@citizen428 yeah, that would definitely be better and will resolve privacy issues. However, eventually the amount of users might increase to more than what the admins can handle. An RSS feed conflict might not occur too frequently though, so I guess this could be feasible.

**Comment by rhymes** [comment]
@squarebat I like @citizen428 suggestion! I don't think the volume of RSS feed conflict is going to be that much. If the volume becomes unbearable, we can always switch to a better alert system for admins. My vote goes to this compromise that preserves privacy and alerts admins.

**Comment by lornajane** [comment]
Is it possible that someone else has added my public feed to their account and this is why can't publish with a canonical URL to dev.to any more? I can't find a live duplicate of my post but I also can't publish with the canonical URL

**Comment by citizen428** [comment]
What's your feed's URL @lornajane?

**Comment by ftisiot** [comment]
I believe I still have this setup @lornajane
